scripts/bfb-install

```diff
@@ -27,19 +27,553 @@
+  wait $pipe_to_rshim_pid
+  exit_code=$?
+  if [ $exit_code -eq 0 ]; then
+    echo "OK"
+  else
+    echo "Warning: pipe_to_rshim process on BMC exited with code $exit_code"
+  fi
+}
+
+# Push the BFB stream to rshim
+#
+# Exit if error.
+#
+push_boot_stream()
+{
+  if [ "$method" == "host_rshim" ]; then
+    push_boot_stream_via_host_rshim
+  elif [ "$method" == "bmc_rshim_scp" ]; then
+    push_boot_stream_via_bmc_rshim_scp
+  elif [ "$method" == "bmc_rshim_nc" ]; then
+    push_boot_stream_via_bmc_rshim_nc
+  elif [ "$method" == "bmc_rshim_ncpipe" ]; then
+    push_boot_stream_via_bmc_rshim_ncpipe
+  else
+    echo "Error: Invalid method: $method"
+    exit 1
+  fi
+}
+
+# Wait for RSHIM to finish updating by monitoring keywords in the RSHIM log
+#
+# Note this function relies on exact behavior of unalias'd echo, so we use
+# "\echo" inside.
+#
+# param 1: mode: host or network
+# param 2: rshim: rshim device name
+# param 3: (Only when mode is network) bmc_ip: BMC host name or IP address
+#
+# Global variables used: None directly
+#
+# Example:
+#  wait_for_update_to_finish host /dev/rshim0
+#  wait_for_update_to_finish network /dev/rshim0 lab105bmc
+wait_for_update_to_finish()
+{
+  local mode=$1
+  local rshim=$2
+  local bmc_ip=$3
+
+  \echo "Collecting BlueField booting status. Press Ctrl+C to stop…"
+
+  # Set display level to 2 to show more information
+  run_cmd_exit $mode "echo 'DISPLAY_LEVEL 2' > ${rshim}/misc"
+
+  last=""
+  finished=0
+  while [ $finished -eq 0 ]; do
+    last_len=${#last}
+    cmd_get_log="cat ${rshim}/misc | sed -n '/^ INFO/,\$p'"
+    cur=$(run_cmd_exit $mode "$cmd_get_log")
+    cur_len=${#cur}
+
+    sleep 1
+
+    if \echo ${cur} | grep -Ei \
+      "Reboot|finished|DPU is ready|In Enhanced NIC mode|Linux up" \
+        >/dev/null; then
+      finished=1
+    fi
+
+    # Overwrite if current length smaller than previous length.
+    if [ ${last_len} -eq 0 -o ${last_len} -gt ${cur_len} ]; then
+        \echo "${cur}" | sed '/^[[:space:]]*$/d'
+      last="${cur}"
+      continue
+    fi
+
+    # Overwrite if first portion doesn't match.
+    sub_cur=$(\echo "${cur}" | dd bs=1 count=${last_len} 2>/dev/null)
+    if [ "${sub_cur}" != "${last}" ]; then
+      \echo "${cur}" | sed '/^[[:space:]]*$/d'
+      last="${cur}"
+      continue
+    fi
+
+    # Nothing if no update.
+    if [ ${last_len} -eq ${cur_len} ]; then
+      [ $finished -eq 0 ] && continue;
+    fi
+
+    # Print the diff.
+    \echo "${cur}" | dd bs=1 skip=${last_len} 2>/dev/null | \
+      sed '/^[[:space:]]*$/d'
+    last="${cur}"
+  done
+}
+
+# Switch to BMC rshim for network rshim methods
+#
+# param: rshim_running_on_host_save: original rshim running status on host.
+#        Values can be 0 or 1.
+# param: rshim_running_on_bmc_save: original rshim running status on BMC. Values
+#        can be 0 or 1.
+#
+# Global variables used: rshim_start_cmd, rshim_stop_cmd
+#
+# Example:
+#   switch_to_bmc_rshim 1 0
+#
```

Comment:
We can't do the automatic switchover because the rshim could be managing multiple rshims. Need to get rid of such related code.